staging/src/k8s.io/apiserver/pkg/authentication/request/x509/x509.go

```diff
@@ -284,10 +285,28 @@ var CommonNameUserConversion = UserConversionFunc(func(chain []*x509.Certificate
 	return &authenticator.Response{
 		User: &user.DefaultInfo{
 			Name:   chain[0].Subject.CommonName,
+			UID:    parseUIDFromCert(chain[0]),
 			Groups: chain[0].Subject.Organization,
 			Extra: map[string][]string{
 				user.CredentialIDKey: {id},
 			},
 		},
 	}, true, nil
 })
+
+// https://datatracker.ietf.org/doc/html/rfc4519#section-2.39
+// http://www.alvestrand.no/objectid/0.9.2342.19200300.100.1.1.html
+var uidOID = asn1.ObjectIdentifier([]int{0, 9, 2342, 19200300, 100, 1, 1})
+
+// Go's crypto/x509 library doesn't explicitly serialize UID to a named field, so
+// we read it off the asn1 array.
+func parseUIDFromCert(cert *x509.Certificate) string {
+	for _, name := range cert.Subject.Names {
+		if name.Type.Equal(uidOID) {
```

Comment:
I assume this `if` statement can only evaluate to `true` for one `name`?  i.e. can a cert have more than one `UID` set?

Reply:
> [2.39](https://datatracker.ietf.org/doc/html/rfc4519#section-2.39).  'uid'
> 
>    The 'uid' ('userid' in [RFC 1274](https://datatracker.ietf.org/doc/html/rfc1274)) attribute type contains computer
   system login names associated with the object.  Each name is one
   value of this multi-valued attribute.
   (Source: [RFC 2798](https://datatracker.ietf.org/doc/html/rfc2798) [[RFC2798](https://datatracker.ietf.org/doc/html/rfc2798)] and [RFC 1274](https://datatracker.ietf.org/doc/html/rfc1274) [[RFC1274](https://datatracker.ietf.org/doc/html/rfc1274)])

Reply:
The OID that applies here is 0.9.2342.19200300.100.1.1 and I think (not sure!) that one is single valued.

Reply:
Very interesting case - thanks everyone for the feedback!

I pushed an update to handle multi-valued `UID`s by selecting the last `UID`. This is consistent with how the `x509/pkix` package handles multi-valued `CN` attribute types: https://cs.opensource.google/go/go/+/master:src/crypto/x509/pkix/pkix.go;l=161?q=FillFromRDNSequence&ss=go%2Fgo
I updated the test to verify behavior when there are multiple `CN` and `UID` attribute types.

Note how `CN` is a `multi-valued` attribute as well:

>   The 'cn' ('commonName' in X.500) attribute type contains names of an
>   object.  Each name is one value of this multi-valued attribute.  If
>   the object corresponds to a person, it is typically the person's full
>   name.

https://datatracker.ietf.org/doc/html/rfc4519#section-2.3

Also note that the order of multi-valued attribute types follow the serialized representation in the ASN.1 structure of X.509 certificate:   https://cs.opensource.google/go/go/+/master:src/crypto/x509/x509.go;l=2142-2149?q=FillFromRDNSequence&ss=go%2Fgo
(ie: they are ordered according to how they were passed into the cert upon creation - there is no additional sorting being done)

Reply:
I'm not sure the `uid` defined in that RFC matches what we mean by user uid

What we mean (the "u" stands for "unique"):
```
	// GetUID returns a unique value for a particular user that will change
	// if the user is removed from the system and another user is added with
	// the same name.
	GetUID() string
```

RFC (the "u" stands for "user"):
```
The 'uid' ('userid' in [RFC 1274](https://datatracker.ietf.org/doc/html/rfc1274)) attribute type contains computer system login names associated with the object.  Each name is one
```

If we were going to use a pre-defined x509 value, `x500UniqueIdentifier` seems much closer to what we mean by `uid`:

https://datatracker.ietf.org/doc/html/rfc4519#section-2.43

> The 'x500UniqueIdentifier' attribute type contains binary strings that are used to distinguish between objects when a distinguished name has been reused.  Each string is one value of this multi-valued attribute.

That said, that seems problematic because it's binary, not text, and it's still multi-valued, and it seems very strange to intentionally map a multi-valued attribute to a single-value in our user interface, and accept certs that specify multiple values for it (if we'd known CN was multi-value, we likely would not have used it for username).

Reply:
I think it's _probably_ safer to obtain / define an oid exactly for our use, to eliminate the chance of any well-behaved CA already issuing certs using a "standard" oid in ways that will cause us issues (putting non-utf8 data in a binary string field, or actually putting multiple values in a multi-value attribute, etc)

Reply:
If we're assigning OIDs, we could perhaps annotate certificates to say "look, for this particular certificate, treat RFC 1274 `userid` as the username, get groups from this multivalue username, and ignore commonName completely".

I'd support a PR that did that other more complex thing. But: complex, it is.

Reply:
I really want to keep backwards compat with existing username / group extraction for n-1 API servers and n-1/2/3 kubelets

Reply:
I was thinking _n+(many)_.

Maybe phase out the existing thing after 10 minor releases? Maybe leave it in even longer behind a flag?
I wish it didn't need a KEP, but it would.

Anyway, I mention the more complex option in case we pick something that makes moving away from `commonName` hard, given we may one day want to.

Reply:
> A new attribute for extracting uid is a strictly additive change we don't have to worry about breaking old servers / kubelets if they encounter a client certificate which contains it

+1 to this approach.  If we want to start messing with `CN`, then that is whole another conversation that will 100% require a KEP and out of scope for this PR.

Reply:
I see a couple approaches for defining a new OID, but not sure how to proceed:
1. It seems [OID 1.3.9900 to 1.3.9999 may be considered as "internal use" OIDs](http://oid-info.com/get/1.3). I pushed an implementation using `1.3.9901.1` as the OID - see the latest commit exemplifying this.
2. We can [apply for a private enterprise number through IANA here](https://www.iana.org/assignments/enterprise-numbers/assignment/apply/). Registration is free, and it takes ~1 week. But who should file it, or which name/address should I use? Happy to help where I can.

(I also posted this question in our related [slack discussions with more details](https://kubernetes.slack.com/archives/C05MW3382US/p1729976457287509?thread_ts=1729791759.941219&cid=C05MW3382US))
Thank you for the feedback :)